src/analysis/processing/qgsalgorithmstdbscanclustering.cpp

```diff
@@ -0,0 +1,339 @@
+                         ---------------------
+    begin                : July 2018
+    copyright            : (C) 2018 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsalgorithmstdbscanclustering.h"
+#include "qgsspatialindexkdbush.h"
+#include <unordered_set>
+
+///@cond PRIVATE
+
+QString QgsStDbscanClusteringAlgorithm::name() const
+{
+  return QStringLiteral( "stdbscanclustering" );
+}
+
+QString QgsStDbscanClusteringAlgorithm::displayName() const
+{
+  return QObject::tr( "ST-DBSCAN clustering" );
+}
+
+QString QgsStDbscanClusteringAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Clusters spatiotemporal point features using a density based scan algorithm." );
+}
+
+QStringList QgsStDbscanClusteringAlgorithm::tags() const
+{
+  return QObject::tr( "clustering,clusters,density,based,points,temporal,time" ).split( ',' );
+}
+
+QString QgsStDbscanClusteringAlgorithm::group() const
+{
+  return QObject::tr( "Vector analysis" );
+}
+
+QString QgsStDbscanClusteringAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectoranalysis" );
+}
+
+void QgsStDbscanClusteringAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ),
+                QObject::tr( "Input layer" ), QList< int >() << QgsProcessing::TypeVectorPoint ) );
+  addParameter( new QgsProcessingParameterField( QStringLiteral( "DATETIME_FIELD" ),
+                QObject::tr( "Date/time field" ), QVariant(),
+                QStringLiteral( "INPUT" ), QgsProcessingParameterField::DateTime, false, false ) );
+
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "MIN_SIZE" ), QObject::tr( "Minimum cluster size" ),
+                QgsProcessingParameterNumber::Integer, 5, false, 1 ) );
+  addParameter( new QgsProcessingParameterDistance( QStringLiteral( "EPS1" ),
+                QObject::tr( "Maximum distance between clustered points" ), 1, QStringLiteral( "INPUT" ), false, 0 ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "EPS2" ),
+                QObject::tr( "Maximum date/time interval (in days unit) between clustered points" ), QgsProcessingParameterNumber::Double, 1, false, 0 ) );
```

Comment:
can you explain the rationale behind using "days" as a unit, vs say hours?

Reply:
Well, the short term rationale is that I felt days as a unit was a good fixed-unit type sitting in between lower units (hours, minutes) and higher units (weeks, months, years).

But it's also a placeholder until I have some free time to code a QgsProcessingParameterInterval, which will have a dropdown combobox to pick the unit type. 

Until then, I think days is tolerable :wink: